### ISSUE TITLE
Need code review

### ISSUE THREAD

**Issue by dtcarney741** [dtcarney741/covid19_data_analysis#1]
Aiden, please review my latest check in for adding the moving average filter functionality.  I'm no sure if I added the GUI selection options correctly.  It seems to work, though sometimes when you select Moving Average Filter, and then switch back to None, the next time you click on Moving Average Filter it doesn't jumps back to None immediately.  It seemed a little convoluted how this function needed to get added, especially with respect to the smooth configuration settings.  We might want to add other smoothing filters in the future, so making this more robust and easier to expand would be a good idea.

**Comment by ajcarney** [comment]
I started working on it and fixed the issues of it switching back to none: the cache was not being set up correctly.

However, now there is a bug when creating a derivative graph where the length of the x and y datasets are not the same size (len(y) is one less than len(x)). I am not sure if this was present before I made any changes. This only occurred after the moving average function was called. I looked in the file where the function is defined and found line 199
```for i in range(1, len(y_data)):```
 meaning the first index will not be included. Will you confirm that this is incorrect and should be 0?

**Comment by dtcarney741** [comment]
I agree that this looks fishy, but it might have been intentional to prevent an indexing of the array below 0, with repsect to the sub_value.  I need to think this through, which I will do later tonight.

**Comment by dtcarney741** [comment]
Ok, I fixed it.  The function was a little off because I over simplified the indexing.  I think this might be able to be done without the extra if statement if you add every pass through the loop, calculate the moving average value and then subtract before the next pass through the loop.  This seems a little clearer to me though as it distinguishes the three cases clearly (1 - before you fill the moving average filter, 2 - the first time the moving average filter is filled and you don't need to subtract an old value, and 3 - every other time after the moving average filter is filled and you need to both add and subtract).

This is how a moving average filter is often implemented in embedded processors because it only requires a single add and subtract instruction per loop.  In PC software that is not resource constrained it could just be accomplished by adding the window_len consecutive values every pass through the loop, and the code wouldn't even need to have any if statements.

Please look it over and make sure you understand how it is working.

**Comment by ajcarney** [comment]
I pushed my changes and the issues should be fixed including the moving average one. I added a new streamlit feature that requires streamlit v0.87.0. Adding new smoothing options should be relatively straight forward. My commit message has a little bit more on how I refactored it
